### PR TITLE
Add KafkaStreams customizers and wrapper configurer for StreamsBuilderFactoryBean.

### PR DESCRIPTION
**Pull request by nyilmaz**
Addresses https://github.com/spring-projects/spring-kafka/issues/621

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@nyilmaz Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-kafka&pullRequestId=625)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-projects/spring-kafka&pullRequestId=625) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@nyilmaz Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-projects/spring-kafka&pullRequestId=625)!

File: spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaStreamsDefaultConfiguration.java [hunk 1]
```diff
@@ -54,10 +63,13 @@
 
 	@Bean(name = DEFAULT_STREAMS_BUILDER_BEAN_NAME)
 	public StreamsBuilderFactoryBean defaultKafkaStreamsBuilder(
-			@Qualifier(DEFAULT_STREAMS_CONFIG_BEAN_NAME) ObjectProvider<StreamsConfig> streamsConfigProvider) {
+			@Qualifier(DEFAULT_STREAMS_CONFIG_BEAN_NAME) ObjectProvider<StreamsConfig> streamsConfigProvider,
+			Optional<List<KafkaStreamsCustomizer>> kafkaStreamsCustomizers) {
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaStreamsDefaultConfiguration.java]: Do we really need here an `Optional` wrapper? 
Won't we have an empty `List` if there is no `KafkaStreamsCustomizer` beans in the context?

On the other hand I don't think that this is good approach if we are going to manage several `KafkaStreams` in the application. This way you are going to consume all the `KafkaStreamsCustomizer`s where you might need just only one.

I wouldn't allow to inject a `List<KafkaStreamsCustomizer>` at all. A single `KafkaStreamsCustomizer` shouldn't be fully enough. And here, on the `@EnableKafkaStreams` level we would use really some conventional bean name do not clash with any other possible `KafkaStreamsCustomizer` in the application context.

File: spring-kafka/src/main/java/org/springframework/kafka/core/KafkaStreamsConfigurer.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.processor.StateRestoreListener;
+
+/**
+ * Wrapper class of configurable properties of {@link KafkaStreams}. Note that
+ * {@code setters} of this configurer has precedence over {@link KafkaStreamsCustomizer}s
+ *
+ * @author Nurettin Yilmaz
+ */
+class KafkaStreamsConfigurer {
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/core/KafkaStreamsConfigurer.java]: I don't see reason in this class. Sounds like `CompositeKafkaStreamsCustomizer` would be better solution.

File: spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java [hunk 3]
```diff
@@ -140,12 +144,21 @@ public void setClientSupplier(KafkaClientSupplier clientSupplier) {
 		this.clientSupplier = clientSupplier; // NOSONAR (sync)
 	}
 
+	public void addKafkaStreamsCustomizers(List<KafkaStreamsCustomizer> kafkaStreamsCustomizers) {
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java]: This must be setter. Remember XML `<bean>`. 
And as I said before: only single `KafkaStreamsCustomizer` should be injected here. The list-like behavior we can reach with the `CompositeKafkaStreamsCustomizer` implementation.

File: spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java [hunk 4]
```diff
@@ -197,22 +209,27 @@ public void stop(Runnable callback) {
...
+				logger.debug(topology.describe());
+				doStart(topology);
 			}
 			catch (Exception e) {
 				throw new KafkaException("Could not start stream: ", e);
 			}
 		}
 	}
 
+	private void doStart(Topology topology) {
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java]: I'm not sure that we need to extract this method. There is no too much complexity and no any exception handling.

File: spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaStreamsDefaultConfiguration.java (hunk 1 above)
Comment (nyilmaz) [reply spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaStreamsDefaultConfiguration.java]: Completely right, fixing.

File: spring-kafka/src/main/java/org/springframework/kafka/core/KafkaStreamsConfigurer.java (hunk 2 above)
Comment (nyilmaz) [reply spring-kafka/src/main/java/org/springframework/kafka/core/KafkaStreamsConfigurer.java]: Could you eloborate on this a little bit? I though that separating settable fields like `stateListener`, `stateRestoreListener` and `uncaughtExceptionHandler` and configure logic from `StreamsBuilderFactoryBean` is more readable.

File: spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java (hunk 4 above)
Comment (nyilmaz) [reply spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java]: Adding
```
Topology topology = getObject().build();
logger.debug(topology.describe());
```
increases its complexity a little.

Comment (artembilan) [reply spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java]: Well, that's great that you took my attention to this `logger.debug()`.
To be honest I don't want to have it, at least not without `logger.isDebugEnabled()`.

You just moved 8 lines of code to the method without any value in the future. 
That new `doStart()` method doesn't bring any distribution of responsibility.
That's how I would like to live without it. 
More over it's not going to be called from any other place.

Right, we might think different way for code design, but our style (at least mine) is slightly different and I'm not convinced yet in this new method.

Thank you for understanding

File: spring-kafka/src/main/java/org/springframework/kafka/core/KafkaStreamsCustomizer.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ */
+
+package org.springframework.kafka.core;
+
+import org.apache.kafka.streams.KafkaStreams;
+
+/**
+ * Callback interface that can be used to configure {@link KafkaStreams} directly.
+ *
+ * @author Nurettin Yilmaz
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/core/KafkaStreamsCustomizer.java]: `@since 2.1.5`

File: spring-kafka/src/main/java/org/springframework/kafka/core/CompositeKafkaStreamsCustomizer.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.kafka.streams.KafkaStreams;
+
+/**
+ * Composite {@code KafkaStreamsCustomizer} customizes {@link KafkaStreams} by delegating.
+ *
+ * @author Nurettin Yilmaz
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/core/CompositeKafkaStreamsCustomizer.java]: `@since 2.1.5`

Also sounds like a wrong package for this API.
Would be better to have it in the `config`, but then we will have a package tangle with the `core` via `StreamsBuilderFactoryBean`.

I think we don't have choice now, but we can reconsider the place for these Kafka Streams related classes in the next `2.2` version.

File: spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaStreamsDefaultConfiguration.java [hunk 7]
```diff
@@ -31,10 +33,14 @@
  * {@link KafkaStreamsDefaultConfiguration#DEFAULT_STREAMS_CONFIG_BEAN_NAME} is present
  * in the application context. Otherwise a {@link UnsatisfiedDependencyException} is thrown.
  *
+ * Also collects {@link KafkaStreamsCustomizer}s to be applied just after {@link KafkaStreams}
+ * created and before {@link KafkaStreams#start()}.
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaStreamsDefaultConfiguration.java]: Sounds like something is missed in this class.
However we discussed with you about some conventional `KafkaStreamsCustomizer`. Similar to what we have now for the `StreamsConfig` with the `DEFAULT_STREAMS_CONFIG_BEAN_NAME`.

Then again: this JavaDoc is not related anymore and has to be changed respectively

File: spring-kafka/src/main/java/org/springframework/kafka/core/CompositeKafkaStreamsCustomizer.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright 2017-2018 the original author or authors.
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/core/CompositeKafkaStreamsCustomizer.java]: This class has been created in this `2018`, so, no any past in the Copyright, please.
Just the actual truth.

File: spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java [hunk 9]
```diff
@@ -140,12 +150,21 @@ public void setClientSupplier(KafkaClientSupplier clientSupplier) {
 		this.clientSupplier = clientSupplier; // NOSONAR (sync)
 	}
 
+	public void setKafkaStreamsCustomizer(KafkaStreamsCustomizer kafkaStreamsCustomizer) {
+		Assert.notNull(kafkaStreamsCustomizer, "'kafkaStreamsCustomizer' must not be null");
+		this.kafkaStreamsCustomizer = kafkaStreamsCustomizer; // NOSONAR (sync)
+	}
+
 	public void setStateListener(KafkaStreams.StateListener stateListener) {
-		this.stateListener = stateListener; // NOSONAR (sync)
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java]: I think you need to reinstate that Sonar related comment.

File: spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java [hunk 10]
```diff
@@ -52,15 +58,19 @@
 
 	private KafkaStreams kafkaStreams;
 
+	private KafkaStreamsCustomizer kafkaStreamsCustomizer = new CompositeKafkaStreamsCustomizer();
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java]: This instance for nothing doesn't have too much value.
Only what it safes for you is just simple `if (this.kafkaStreamsCustomizer != null)`.
I'd prefer such an `if`, instead of extra method call.

File: spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java [hunk 11]
```diff
@@ -198,9 +216,15 @@ public synchronized void start() {
...
+				Topology topology = getObject().build();
+				if (logger.isDebugEnabled()) {
+					logger.debug(topology.describe());
+				}
+				this.kafkaStreams = new KafkaStreams(topology, this.streamsConfig, this.clientSupplier);
 				this.kafkaStreams.setStateListener(this.stateListener);
-				this.kafkaStreams.setUncaughtExceptionHandler(this.exceptionHandler);
+				this.kafkaStreams.setGlobalStateRestoreListener(this.stateRestoreListener);
+				this.kafkaStreams.setUncaughtExceptionHandler(this.uncaughtExceptionHandler);
+				this.kafkaStreamsCustomizer.customize(this.kafkaStreams);
```
Comment (nyilmaz) [inline spring-kafka/src/main/java/org/springframework/kafka/core/StreamsBuilderFactoryBean.java]: @artembilan here one may be overwriting `stateRestoreListener`, `uncaughtExceptionHandler` or `stateListener` via customizer api. Do you think we should warn users about that?

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-kafka/commit/e77e1efbc5ef9c27d6aa50d8a7c1e9fd9962b516.

@nyilmaz ,

Thank you for contribution.
Looking forward for more!
